### ISSUE TITLE
if condition with incorrect parameters causes curious errors, expected to find )

### ISSUE THREAD

**Issue by leonschwanitz** [dart-lang/sdk#46164]
System Informations:
Dart SDK Version: 2.12.3 (Stable) on Windows_x64 using Flutter 2.0.5

When we use the following code:
```dart
if (!data[i]==null or !data[i]=='' ){

}
```

#### The or in the if statement is wrong here in this case and should be replaced with ||. But the compiler in VS Code does not show this in this column, but says at the top of the file that a bracket is missing. "Expected to find ')'. dart(expected_token)".

I think it would be better if dart would indicate directly in the line that something is wrong with the if condition to save the long search. :)

**Comment by vsmenon** [comment]
Can you distill this to a small file?  Is it possible you have another error somewhere in the code?

On 2.13.1 Stable, I'm seeing a good error message, and, in VSCode, it takes me to the line in question:

![Screen Shot 2021-05-27 at 6 11 46 AM](https://user-images.githubusercontent.com/2119553/119832028-ab295180-beb2-11eb-9650-6198069db985.png)

**Comment by jensjoha** [comment]
Thank you for reporting this.
Writing out "or" has specifically been handled though. If I try to run this code today (on a new compile):
```
void main() {
  int i = 0;
  List data = [null];
  if (!data[i]==null or !data[i]==''){
  }
}
```

I get this output:
```
t.dart:4:22: Error: Binary operator 'or' is written as '||' instead of the written out word.
Try replacing 'or' with '||'.
  if (!data[i]==null or !data[i]==''){
                     ^^
```

While I'm not sure when this made it in exactly, using the dart included in Flutter 2.2 does the same thing.

**Comment by jensjoha** [comment]
The thing is the flutter version (`Flutter 2.0.5`) which included a dart version (`2.12.3`) with a parser bug in it. It was fixed in 16a66c89025ad330844a07baf76553c45cfbf7b8 which is not a part of 2.12.3.
The new flutter stable (`2.2.0`) includes a newer dart version (`2.13.0`) which does include the fix.
